### PR TITLE
[tmva][sofie] Fix Conv/Pool padding & ceil_mode bugs, and harden the SOFIE test suite

### LINKED ISSUE #21873: [tmva][sofie] Extend Im2col to support asymmetric padding for full ONNX SAME autopad compliance
### Explain what you would like to see improved and how.

While working on PR #21793 (fixing SAME_UPPER/SAME_LOWER autopad formula), I noticed that `UTILITY::Im2col` in `SOFIE_common.hxx` only supports symmetric padding. The function signature takes a single `pad_h` and `pad_w` per dimension and internally uses `2 * pad_h` to compute the output size:

```cpp
output_h = (height + 2 * pad_h - (dilation_h * (kernel_h - 1) + 1)) / stride_h + 1;
```
This works fine for odd kernels with SAME padding because total padding is always even, so `begin == end`. But for even kernels (e.g. kernel=2, input=4, stride=1), ONNX SAME_UPPER gives totalPad=1, i.e. pad_begin=0 and pad_end=1. Im2col only receives the begin pad and doubles it, so it computes the wrong output size and wrong start position.
The Conv operator's shape inference already handles this correctly using fAttrPads[0] + fAttrPads[i1] (begin + end), but Im2col is inconsistent with that.

The fix would be extending Im2col (and Im2col_3d) to accept separate begin and end pads per dimension, and updating callers in ROperator_Conv.hxx accordingly.

**Note: However I think this is low priority since mainstream networks use odd kernels, but it would make SOFIE fully ONNX-compliant for SAME autopad across all kernel sizes.**

### How it could be improved
- Change Im2col signature to accept `pad_h_begin`, `pad_h_end`, `pad_w_begin`, `pad_w_end`
- Update output size formula to use `pad_h_begin + pad_h_end` instead of `2 * pad_h`
- Update start offset to use `-pad_h_begin`
- Update all callers in ROperator_Conv.hxx and ROperator_Pool.hxx

### ROOT version

6.39.01

### Installation method

Built from source

### Operating system

Linux (Ubuntu 22.04)

### Additional context

_No response_

### LINKED ISSUE #22523: [tmva][sofie] AveragePool with ceil_mode divides the partial/overhang window by the full kernel size
### Check duplicate issues.

- [x] Checked for duplicates

### Description

When an ONNX AveragePool node has ceil_mode=1, the last sliding window can extend past the end of the input (the "overhang" window that ceil_mode is meant to keep). SOFIE's generated code averages such a window by dividing the sum of the real cells by the full kernel size, instead of by the number of cells actually present. The result is wrong for that window.

The divisor in `ROperator_Pool.hxx` is computed at run time only when `count_include_pad == 0 && doPadding`. Otherwise it is the compile-time constant kernel area (`kh`, `kw*kh`, ...). A window can also be partial because of ceil_mode, which this condition does not cover, so:

- count_include_pad = 0 (the default) with no padding + ceil_mode -> constant divisor, wrong.
- count_include_pad = 1 + ceil_mode -> constant divisor includes the overhang cells, wrong.

Example (1D): input [1,2,3,4,5,6], kernel 3, stride 2, no padding, ceil_mode = 1. Output has 3 windows; the third covers input indices [4,5] (index 6 is the overhang).

- Expected (PyTorch avg_pool1d, count_include_pad=False; same as the ONNX definition):  [2.0, 4.0, 5.5]   ((5+6)/2 = 5.5)
- SOFIE generated code: [2.0, 4.0, 3.6667] ((5+6)/3 = 3.6667)

The first two windows are full so they match; only the overhung window is wrong.

### Reproducer

Three small steps to reproduce- 

1. Create the model (Python, needs the `onnx` package):
```
##make_avgpool_ceil.py
import onnx
from onnx import helper, TensorProto
node = helper.make_node('AveragePool', ['x'], ['y'],
                        kernel_shape=[3], strides=[2], ceil_mode=1)
g = helper.make_graph([node], 'avgpool_ceil',
    [helper.make_tensor_value_info('x', TensorProto.FLOAT, [1, 1, 6])],
    [helper.make_tensor_value_info('y', TensorProto.FLOAT, [1, 1, 3])])
m = helper.make_model(g, opset_imports=[helper.make_opsetid('', 14)])
m.ir_version = 7
onnx.save(m, 'avgpool_ceil.onnx')
```
   Run: python3 make_avgpool_ceil.py

2. Generate the SOFIE code (gen.C):
```
// gen.C
void gen() {
   using namespace TMVA::Experimental::SOFIE;
   RModelParser_ONNX parser;
   RModel model = parser.Parse("avgpool_ceil.onnx");
   model.Generate();
   model.OutputGenerated("avgpool_ceil.hxx");
}
```
   Run: root -l -b -q gen.C

3. Run inference and print the output (run.C):
```
// run.C
#include "avgpool_ceil.hxx"
void run() {
   std::vector<float> x = {1, 2, 3, 4, 5, 6};
   TMVA_SOFIE_avgpool_ceil::Session s;
   auto y = s.infer(x.data());
   for (float v : y) std::cout << v << " ";
   std::cout << std::endl;
}
```
   Run: root -l -b -q run.C
   Prints: 2 4 3.66667     (expected: 2 4 5.5)

(If the generated namespace differs, check the top of avgpool_ceil.hxx and adjust
TMVA_SOFIE_avgpool_ceil accordingly.)

### ROOT version

6.41.01

### Installation method

Built from source

### Operating system

Ubuntu 22.04.2 LTS

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by guitargeek**
**Closes https://github.com/root-project/root/issues/22523**
**Closes https://github.com/root-project/root/issues/21873**

---------------------

This PR fixes a set of correctness bugs in the SOFIE `Conv` and `Pool` operators around padding and `ceil_mode`, and in the process substantially hardens the SOFIE ONNX test suite. Several tests turned out to pass even against a broken implementation.

## Operator fixes

- **AveragePool divisor for partial `ceil_mode` windows** (`d4d8792`, closes #22523). The divisor was only computed at run time for `count_include_pad == 0 && doPadding`; a window made partial by `ceil_mode` (overhanging the input) was wrongly averaged over the full kernel area. The divisor is now the kernel extent clipped to the input whenever there is padding or `ceil_mode`, and stays the `constexpr` kernel area when every window is full (so ordinary pooling generates identical code). E.g. input `[1..6]`, kernel 3, stride 2, `ceil_mode 1` now returns `[2, 4, 5.5]` instead of `[2, 4, 3.6667]`.

- **Ignore `ceil_mode` windows that start past the input** (`a756e86`). Rounding the output dimension up could add a sliding window starting entirely inside the right padding; ONNX drops such a window but SOFIE emitted it, so both the inferred shape and the generated loop had one window too many. Now clipped in `ShapeInference` and in `Generate()`, only when `ceil_mode` is set (no-op in floor mode). Verified against onnxruntime over a sweep of 1760 1D pooling configurations. Affects both `MaxPool` and `AveragePool`.

- **Asymmetric padding in `Im2col`** (`e8c5144`, closes #21873). `Conv`'s shape inference already handled per-end padding (`fAttrPads[d] + fAttrPads[d + fDim]`), but `Im2col`/`Im2col_3d` collapsed the two pads to their average. Because the `_xcol` buffer is sized from the correctly-inferred output shape, this wrote a different number of columns than were allocated and the convolution then read uninitialized memory (e.g. `SAME_UPPER` on a 1×1×4×4 input with a 2×2 kernel gave garbage). `Im2col`/`Im2col_3d` now take begin/end pads separately. Adds six models covering the 1d/2d/3d/grouped paths with `SAME_UPPER`, `SAME_LOWER` and explicit `pads`, and re-enables `ConvWithAsymmetricPadding` (which was failing for an unrelated wrong declared-output-shape).

## Test-suite hardening

- **Fix tests with non-discriminating reference values** (`4518137`). Several tests passed against a broken implementation: `LinearWithLeakyRelu` used a tolerance of 1 (all-zeros passed 23/24 elements); `Concat0D` compared its size to itself; `Greater`/`LessOrEqual` shared a non-discriminating input pair; `Sub`'s reference was symmetric. All now use discriminating references and the standard `DEFAULT_TOLERANCE`.

- **Use the comparison helpers instead of hand-rolled loops** (`5772ee2`). ~30 tests re-implemented `expectNear()`/`expectEqual()` less safely: checking size with `EXPECT_EQ` and then bounding the loop by `output.size()`, so an oversized output read past the reference array instead of failing. Adds `vector<vector<T>>` overloads for the multi-output models (`Split`, `Clip`). No coverage change: all 153 tests still pass.

- **Share the common code of the PyTorch model generators** (`e97ccc2`). The six generators were copy-pasted and had drifted; the `--bn`/`--maxpool`/`--avgpool` paths were broken in three of the four convolution generators (e.g. `Conv1d` using `BatchNorm2d`/`MaxPool2d`). Command-line parsing, ONNX export and expected-output writing move to a shared `ModelGeneratorUtils.py`, each generator keeps only its network and input tensor, and the dimensionality-correct layers are used throughout.

## Testing

All SOFIE tests pass. The new operator tests each fail before their respective fix and pass after, and perturbing any reference value still fails the corresponding test.

### CODE REVIEW

**Review by dpiparo — APPROVED** [review]
Thanks. This completes the "SOFIE Sprint"
